### PR TITLE
Fix memory leak issue in `torch_fx` tests

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

~~**Question**: On GPU VMs, we have to use `spawn`, see [here](https://github.com/huggingface/transformers/pull/18547#issuecomment-1210375716). However, it still hangs with `spawn` (I can't figure out this yet). Should we have 2 branches: one using new process for CPU VM (on CircleCI), and another one using the original approach (no new process - for GPU VM, like on scheduled CI)?~~

**I might have a solution!** --> send the model to the child process in CPU and send to CUDA device there.

~I am going to try `torch.multiprocessing` first.~ not working neither

----

Run torch_fx tests in a spawn process to avoid [memory issue](https://github.com/huggingface/transformers/issues/18525#issue-1331914135).

- See [this comment](https://github.com/huggingface/transformers/pull/18547#issuecomment-1210260525) for the effect
- The reason to use `JoinableQueue` instead of `Queue` for the outputs:

https://discuss.pytorch.org/t/using-torch-tensor-over-multiprocessing-queue-process-fails/2847

### CODE REVIEW

**Comment by ydshieh** [comment]
- without new process
  - 30 minutes for 100 runs
  - 15 MB leak per run

- with `fork`
  - 5 minutes for 100 runs
  - 1 MB leak per run
  - hangs if `MKL_NUM_THREADS` > 1

- with `spawn`
  - 2~3 minutes for 100 runs
  - 1 MB leak per run

**Comment by ydshieh** [comment]
When , using the new process approach, in some cases, setting `ulimit -n 2048` is necessary.

Otherwise, we might get the following error:
```bash
tests/models/bart/test_modeling_bart.py::BartModelTest::test_torch_fx Traceback (most recent call last):
  File "/usr/lib/python3.9/multiprocessing/queues.py", line 245, in _feed
  File "/usr/lib/python3.9/multiprocessing/reduction.py", line 51, in dumps
  File "/home/yih_dar_huggingface_co/.local/lib/python3.9/site-packages/torch/multiprocessing/reductions.py", line 358, in reduce_storage
RuntimeError: unable to open shared memory object </torch_46201_690006289_939> in read-write mode: Too many open files (24)
```

More details:
```bash
>   ???

tests/test_modeling_common.py:769: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_modeling_common.py:866: in _create_and_check_torch_fx_tracing
    ???
/usr/lib/python3.9/multiprocessing/process.py:121: in start
    ???
/usr/lib/python3.9/multiprocessing/context.py:277: in _Popen
    ???
/usr/lib/python3.9/multiprocessing/popen_fork.py:19: in __init__
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <multiprocessing.popen_fork.Popen object at 0x7fa12a499820>, process_obj = <ForkProcess name='ForkProcess-10' parent=46201 initial>

>   ???
E   OSError: [Errno 24] Too many open files

/usr/lib/python3.9/multiprocessing/popen_fork.py:64: OSError
```

**Comment by ydshieh** [comment]
With GPU, we have to use `spawn`, otherwise

```
Process ForkProcess-1:
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/transformers/tests/test_modeling_common.py", line 143, in _run_torch_jit
    model, input_names, filtered_inputs = in_queue.get(timeout=30)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 116, in get
    return _ForkingPickler.loads(res)
  File "/usr/local/lib/python3.8/dist-packages/torch/multiprocessing/reductions.py", line 112, in rebuild_cuda_tensor
    torch.cuda._lazy_init()
  File "/usr/local/lib/python3.8/dist-packages/torch/cuda/__init__.py", line 207, in _lazy_init
    raise RuntimeError(
RuntimeError: Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method
```

File: tests/test_modeling_common.py
Comment (ydshieh) [inline tests/test_modeling_common.py:826]: not working with GPU/CUDA - need to be `spawn` in this case.

File: tests/test_modeling_common.py
Comment (ydshieh) [inline tests/test_modeling_common.py:835]: This still hangs on GPU VM, even with `spawn`

File: tests/test_modeling_common.py
Comment (ydshieh) [inline tests/test_modeling_common.py:838]: Don't know the exact reason, but passing model in CUDA via Queue will cause tracing issue. Let's pass it in CPU and send to CUDA device in the subprocess.

Comment (sgugger) [reply tests/test_modeling_common.py:838]: You cannot really use multiprocessing with CUDA, they don't play along well together.

**Review by sgugger — COMMENTED** [review]
@michaelbenayoun would be the best person to review this :-)

**Comment by michaelbenayoun** [comment]
I think it's safe to only run those tests on CPU.
Also, when running locally it takes ~ 1 min (althought I agree my machine might be more powerful).

File: tests/test_modeling_common.py
Comment (michaelbenayoun) [inline tests/test_modeling_common.py:141]: What I don't "like" here is that now a test would fail for torchscripting before failing for output mismatch.

File: tests/test_modeling_common.py
Comment (michaelbenayoun) [inline tests/test_modeling_common.py:823]: How does it behave locally?

**Review by michaelbenayoun — COMMENTED** [review]
Left a few comments.
Also be careful because some models also have their "custom" implementation of the test.

File: tests/test_modeling_common.py
Comment (ydshieh) [reply tests/test_modeling_common.py:141]: I agree. I could even running the match in this function, return the matching results to the parent process, and fail the tests there - if you prefer.

File: tests/test_modeling_common.py
Comment (ydshieh) [reply tests/test_modeling_common.py:823]: On CircleCI, we have `MKL_NUM_THREADS = 1` (and we are on CPU) --> no issue.
On scheduled CI, we have `MKL_NUM_THREADS = 8`, but the device is GPU, so we use `spawn` --> no issue.

Locally, it seems that if we don't set `MKL_NUM_THREADS` explicitly, it will behave like `MKL_NUM_THREADS > 1`, and hangs when running on CPU. --> This is not desirable, I will try if I can set it to `1` temporarily inside this test.

File: tests/test_modeling_common.py
Comment (ydshieh) [reply tests/test_modeling_common.py:141]: > I think it's safe to only run those tests on CPU. Also, when running locally it takes ~ 1 min (althought I agree my machine might be more powerful).

Do you run against my branch?

Comment (michaelbenayoun) [reply tests/test_modeling_common.py:141]: Yes it would be perfect!
I did not run anything on your branch.

Comment (ydshieh) [reply tests/test_modeling_common.py:141]: It is indeed fast. This PR is not addressing the time issue, but the memory issue. Currently, each call to `test_torch_fx` will increase the memory usage by ~15MB.

The time issue comes after the fix, as we create new processes to run (part of) the code. Using `fork` is fine, but `spawn` will be quite slow. But currently, this time issue is insignificant (and `spawn` is only used on GPU CI running on our own runners - therefore no a real constraint).

**Comment by ydshieh** [comment]
@michaelbenayoun 
 
I move (almost) the whole testing logic to the child process. On more advantage here is to create the model in the child process, so we don't need to pass it between the process.

Now running 100 times, we have only `0.15 MB increase of memory usage`.

**Comment by ydshieh** [comment]
@michaelbenayoun You are right, some model overwrites `_create_and_check_torch_fx_tracing`. This won't fail this PR however, those models with just run the `test_torch_fx*` tests in the current manner (i.e. not in the child process). I will take a look if those overwritting are necessary. In any case, we can merge this PR as it is (if you are happy with it), and work on those models later.

**Comment by michaelbenayoun** [comment]
I think it's okay now with the changes you've made!

**Comment by ydshieh** [comment]
> I think it's okay now with the changes you've made!

Would love to have a approval from you, @michaelbenayoun.
But no need to rush - as long as you finally happy with the change and click the button.

**Review by michaelbenayoun — APPROVED** [review]

**Comment by ydshieh** [comment]
ready for @sgugger and/or @LysandreJik to have a final check 🚀

File: tests/test_modeling_common.py
Comment (LysandreJik) [inline tests/test_modeling_common.py:142]: Should this be at the top of the file?

File: tests/test_modeling_common.py
Comment (LysandreJik) [inline tests/test_modeling_common.py:160]:
```suggestion
            # FSTM still requires this hack -> FSTM should probably be refactored similar to BART afterward
            model.config.use_cache = False
```

**Review by LysandreJik — APPROVED** [review]
Ok! Thanks for working on this, @ydshieh

File: tests/test_modeling_common.py
Comment (LysandreJik) [reply tests/test_modeling_common.py:142]: sounds good

**Comment by ydshieh** [comment]
I will merge this afternoon, after adding a short command in `_create_and_check_torch_fx_tracing` explaining why we need this change, with a link to #18525

**Comment by ydshieh** [comment]
Hi @michaelbenayoun, I just saw that I fixed a similar issue a few months ago

https://github.com/ydshieh/transformers/blob/489e80cd11b5d2cc21628a89e8fc842de14b906b/tests/test_modeling_common.py#L871

(for `_create_and_check_torchscript`). I am going to change this PR to simply apply that fix. Is it OK for you?

**Comment by ydshieh** [comment]
Changed the PR to simply call `clear_torch_jit_class_registry`. Test failure is irrelevant to this PR - merge now.

**Pull request merged**
